File: Leetcode0057.java
package com.y2020.m01.d02;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * Created by zhangweilong on 2020/01/02
 *
 * @author zhangweilong@corp.netease.com
 * @since 2020/01/02
 */
public class Leetcode0057 {
    public static void main(String[] args) {
        Leetcode0057 sol = new Leetcode0057();
        System.out.println(Arrays.deepToString(sol.insert(new int[][]{
                {1, 3},
                {6, 9},
        }, new int[]{2, 5})));
    }
    public int[][] insert(int[][] intervals, int[] newInterval) {
        if (intervals == null || intervals.length == 0) return new int[][]{newInterval};
        List<int[]> res = new LinkedList<>();
        int i = 0;
        while (i < intervals.length && intervals[i][1] < newInterval[0]) {
            res.add(intervals[i]);
            i++;
        }
        int left = newInterval[0];
        int right = newInterval[1];
        while (i < intervals.length && intervals[i][0] <= newInterval[1]) {
            left = Math.min(left, intervals[i][0]);
            right = Math.max(right, intervals[i][1]);
            i++;
        }
        res.add(new int[]{left, right});
        while (i < intervals.length) {
            res.add(intervals[i++]);
        }
        int[][] r = new int[res.size()][2];
        for (int k = 0; k < res.size(); k++) {
            r[k] = res.get(k);
        }
        return r;
    }
}

File: Leetcode0057.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& a, vector<int>& b) {
        vector<vector<int>> res;
        int k = 0, n = a.size();
        //先把b区间左边和b没有交集的加入到答案里
        while (k < n && a[k][1] < b[0]) res.push_back(a[k++]);
        //合并有交集的区间，并加入到答案里
        if (k < n) {
            b[0] = min(b[0], a[k][0]);
            while (k < n && a[k][0] <= b[1]) b[1] = max(b[1], a[k++][1]);
        }
        res.push_back(b);
        //把在b区间右边和b没有交集的加入到答案里
        while (k < n) res.push_back(a[k++]);
        return res;
    }
};
